pkg/storage/fs/ocis/upload.go

```diff
@@ -44,15 +45,28 @@ var defaultFilePerm = os.FileMode(0664)
 func (fs *ocisfs) Upload(ctx context.Context, ref *provider.Reference, r io.ReadCloser) (err error) {
 	upload, err := fs.GetUpload(ctx, ref.GetPath())
 	if err != nil {
-		return errors.Wrap(err, "ocisfs: error retrieving upload")
+		// Upload corresponding to this ID was not found.
+		// Assume that this corresponds to the resource path to which the file has to be uploaded.
+		buf := &bytes.Buffer{}
+		length, err := io.Copy(buf, r)
```

Comment:
this buffers the whole file in memory before writing it to disk.

Reply:
I think it makes more sense to revert the changes and use the lookup of the node before ... I can look into making the chunking work.

Reply:
and this reads consumes the body, which makes the subsequent writeChunk not read anything ...

Reply:
@butonic the refactoring concerns less the chunking part and more that we can support both TUS and non-TUS uploads out of the box without ugly config changes.

This step was just to get the length of the upload since that is not explicitly provided. One way to get around this is to set `SizeIsDeferred` to true

Reply:
Yes ... and `Upload()` was implementing the non tus based upload. For tus you need `InitiateUpload` and `UseIn`, which you correctly check in the dataprovider ... why did you need to touch the `Upload` implementation?

Reply:
The latest commit gets rid of the buffer issue.

The aim was to have uniform implementation for all types of uploads. Calling InitiateUpload for TUS uploads and not doing so for others didn't seem ideal. Also, since all the file systems did in fact have the `UseIn` method, the PUT calls were handled by a wrapper around the tus client which was also a hack. https://github.com/cs3org/reva/blob/ad43bae0ce75a093eb3915c5414a99f7bac7cc87/internal/http/services/dataprovider/put.go#L59

We wanted to add support for multiple protocols as started in https://github.com/cs3org/reva/pull/828, so all of these should ideally have the same workflow: Call InitiateUpload -> Send PUT/POST call to the returned URL

Reply:
testing the PR